### PR TITLE
Add variant filtering

### LINKED ISSUE #13: Allow opting out of repo addition
Right now we add the r8 repo automatically. To give users back control, it would be nice if we gave them the ability to opt out and define their own

### LINKED ISSUE #15: Add variant filtering API
This should allow for more fine-grained control over which variants get generation, vs just depending on build types like we do today

### PR DESCRIPTION
**Pull request by ZacSweers**
Resolves #15 

Now you can do something like this to explicitly filter out certain variants

```gradle
keeper {
  variantFilter {
    setIgnored(name == "variantThatShouldTotallyBeIgnored")
  }
}
```

~~This works based on eyeballing, now just need to add a test to lock it down~~

### CODE REVIEW

File: docs/configuration.md [hunk 1]
```diff
@@ -69,7 +69,29 @@ if (!hasProperty("productionBuild")) {
 }
 ```
 
-##### <your build here>
+### Variant filter
+
+You can specify a `variantFilter` on the `keeper` extension and dynamically configure which variants
```
Comment (jschear) [inline docs/configuration.md]: Accidentally included the start of this paragraph twice?

File: keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt [hunk 2]
```diff
@@ -258,3 +277,29 @@ internal fun String.capitalize(locale: Locale): String {
...
+private fun unwrapTestedVariant(variantData: Any?): BaseVariant? {
+  return when (variantData) {
+    is BaseVariant -> {
+      when (variantData) {
+        is TestVariant -> variantData.testedVariant
+        is UnitTestVariant -> variantData.testedVariant as? BaseVariant
+        else -> variantData
+      }
+    }
+    is KaptVariantData<*> -> unwrapTestedVariant(variantData.variantData)
```
Comment (jschear) [inline keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt]: `appExtension.testVariants` only contains `TestVariant`s (which `KaptVariantData` doesn't implement), right? I'm not following why this is necessary -- is this an AGP 4.0.0 thing?

File: keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt [hunk 3]
```diff
@@ -127,7 +131,22 @@ class KeeperPlugin : Plugin<Project> {
         val androidJarRegularFileProvider = project.layout.file(androidJarFileProvider)
 
         appExtension.testVariants.configureEach {
```
Comment (jschear) [inline keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt]: nit: the `matching { ... }` API here could be nice to separate the filtering of `TestVariant`s from their configuration. Up to you.

File: keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt (hunk 2 above)
Comment (ZacSweers) [reply keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt]: Ahh good call. Copy-pasta from the redacted plugin code I borrowed this from

File: keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt (hunk 3 above)
Comment (ZacSweers) [reply keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt]: I thought about it but it comes at the cost of possibly unwrapping twice, so I just left it as-is

File: docs/configuration.md (hunk 1 above)
Comment (ZacSweers) [reply docs/configuration.md]: 41c9252

File: keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt (hunk 2 above)
Comment (ZacSweers) [reply keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt]: 0bcc473

File: keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt [hunk 4]
```diff
@@ -127,7 +130,22 @@ class KeeperPlugin : Plugin<Project> {
...
-          val appVariant = testedVariant
+          val appVariant = unwrapTestedVariant(this)
```
Comment (jschear) [inline keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt]: I think you can go back to accessing `testedVariant` directly here. `UnitTestVariant` isn't a subclass/interface of `TestVariant`.

Comment (ZacSweers) [reply keeper-gradle-plugin/src/main/kotlin/com/slack/keeper/KeeperPlugin.kt]: 3d61945

**Pull request merged**